os/kernel/fault_manager/fault_manager.c

```diff
@@ -0,0 +1,291 @@
+ * Name: kill_app
+ *
+ * Description:
+ *   This function will kill all the task/thread created by the root application
+ *   i.e input pid.
+ *   If any fault occurs in the system, fault manager will call this function to
+ *   kill all the thread/task created by the input application.
+ *
+ * Input parameters:
+ *   pid   -   The pid of the applciation, whoes all childs to be killed
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ ****************************************************************************/
+int kill_app(pid_t pid)
+{
+	int ret;
+
+	/* TODO: Implementation of traverser of tree and kill all the task/thread
+	 * of the root application
+	 */
+	ret = kill(pid, SIGWORK);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: fault_manager
+ *
+ * Description:
+ *   This function will receive the faulty application's pid and check if it is one
+ *   of the registered application with fault manager.
+ *   If the application is registered then it will be restarted after killing its
+ *   child processes, and if it is not registered then board will be rebooted.
+ *
+ * Input parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ * Assumptions:
+ *    All the applications of the system will be registered with the fault manager
+ *
+ ****************************************************************************/
+int fault_manager(int argc, char *argv[])
+{
+	mqd_t mqdes;
+	int ret;
+	int i;
+	int index = -1;
+	pid_t pid;
+	pid_t pid_restart;
+	pid_t app_id;
+	struct mq_attr attr;
+	char restart_app[PATH_LEN] = " ";
+
+	attr.mq_maxmsg = 32;
+	attr.mq_msgsize = sizeof(pid);
+	attr.mq_flags = 0;
+
+	mqdes = mq_open(FAULT_MGR_MQ, O_CREAT | O_RDONLY, 0666, &attr);
+	DEBUGASSERT(mqdes != (mqd_t)-1);
+
+	while (1) {
+		sleep(2);
+
+		/* Wait for the message from assert handler for the faulty app pid */
+		ret = mq_receive(mqdes, (char *)&pid, sizeof(pid), NULL);
+		DEBUGASSERT(ret != -1);
+		DEBUGASSERT(ret == sizeof(pid));
+
+		/* Get the root app pid from the pid got from assert handler */
+		app_id = get_appid(pid);
+		DEBUGASSERT(app_id != ERROR);
+
+		fmlldbg("faulty app pid = %d\n", app_id);
+
+		for (i = 0; i < CONFIG_NUM_APPS; i++) {
+			if (reg_table[i].appid == app_id) {
+				index = i;
+				break;
+			}
+		}
+
+		/* Check if the application is registered with fault manager */
+		if (index >= 0 && index < CONFIG_NUM_APPS) {
+			strncpy(restart_app, reg_table[index].app_path, strnlen(reg_table[index].app_path, PATH_LEN) + 1);
+
+			/* Kill the root application and all its child processes */
+			ret = kill_app(app_id);
+			DEBUGASSERT(ret != ERROR);
+
+			/* Deregister the application from fault manager */
+			ret = fm_deregister_app(reg_table[index].appid);
+			DEBUGASSERT(ret != ERROR);
+
+			index = -1;
+			sleep(2);
```

Comment:
Ditto, why sleep 2 here again??

Reply:
Firstly, I added this sleep to add some delay between killing the application and restarting it again, I though it is better to have some delay before restarting it.
Secondly, If application is going to continuous asserts then this sleep will provide other applications to get time to schedule.
What is your opinion on this?